targets/TARGET_ARM_FM/TARGET_FVP_MPS2/us_ticker.c

```diff
@@ -16,77 +16,96 @@
+    US_TICKER_TIMER1->TimerControl  = 0x0ul; // disable TIMER1 and reset all control
+    US_TICKER_TIMER2->TimerControl  = 0x0ul; // disable TIMER2 and reset all control
+
+    US_TICKER_TIMER1->TimerLoad     = 0xFFFFFFFFul;
+    US_TICKER_TIMER2->TimerLoad     = 0xFFFFFFFFul;
+
+    US_TICKER_TIMER1->TimerControl |= CMSDK_DUALTIMER1_CTRL_SIZE_Msk; // set TIMER1 to 32 bit counter
+    US_TICKER_TIMER2->TimerControl |= CMSDK_DUALTIMER2_CTRL_SIZE_Msk; // set TIMER2 to 32 bit counter
 
-    US_TICKER_TIMER1->TimerControl = 0x62; // enable interrupt and set to 32 bit counter and set to periodic mode
-    US_TICKER_TIMER2->TimerControl = 0x42; // enable interrupt and set to 32 bit counter
+    US_TICKER_TIMER1->TimerControl |= 0x1 << CMSDK_DUALTIMER1_CTRL_PRESCALE_Pos; // set TIMER1 with 4 stages prescale
+    US_TICKER_TIMER2->TimerControl |= 0x1 << CMSDK_DUALTIMER2_CTRL_PRESCALE_Pos; // set TIMER2 with 4 stages prescale
 
-    US_TICKER_TIMER1->TimerControl |= 0x80; // enable counter
-    US_TICKER_TIMER2->TimerControl |= 0x80; // enable counter
+    US_TICKER_TIMER2->TimerControl |= CMSDK_DUALTIMER2_CTRL_MODE_Msk;     // set TIMER2 periodic mode
+
+    US_TICKER_TIMER2->TimerControl |= CMSDK_DUALTIMER2_CTRL_ONESHOOT_Msk; // set TIMER2 one-shot mode
+
+    US_TICKER_TIMER1->TimerControl |= CMSDK_DUALTIMER1_CTRL_EN_Msk; // enable TIMER1 counter
+    US_TICKER_TIMER2->TimerControl |= CMSDK_DUALTIMER2_CTRL_EN_Msk; // enable TIMER2 counter
 
     NVIC_SetVector(US_TICKER_TIMER_IRQn, (uint32_t)us_ticker_irq_handler);
-    NVIC_EnableIRQ(US_TICKER_TIMER_IRQn);
+    us_ticker_inited = 1;
+}
+
+void us_ticker_free(void)
+{
+    US_TICKER_TIMER1->TimerControl  &= ~CMSDK_DUALTIMER1_CTRL_EN_Msk; // disable TIMER1
+    US_TICKER_TIMER2->TimerControl  &= ~CMSDK_DUALTIMER2_CTRL_EN_Msk; // disable TIMER2
+    us_ticker_disable_interrupt();
+    us_ticker_inited = 0;
 }
 
 uint32_t us_ticker_read()
 {
-    uint32_t return_value = 0;
-    if (!us_ticker_inited) {
-        us_ticker_init();
-    }
-    return_value = ((~US_TICKER_TIMER2->TimerValue) / 25);
-    return return_value;
+    return ~US_TICKER_TIMER1->TimerValue;
 }
 
 void us_ticker_set_interrupt(timestamp_t timestamp)
 {
-    if (!us_ticker_inited) {
-        us_ticker_init();
-    }
-
-    uint32_t delta = timestamp - us_ticker_read();
-    // enable interrupt
-    US_TICKER_TIMER1->TimerControl = 0x0; // disable timer
-    US_TICKER_TIMER1->TimerControl = 0x62; // enable interrupt and set to 32 bit counter and set to periodic mode
-    US_TICKER_TIMER1->TimerLoad = (delta) * 25; //initialise the timer value
-    US_TICKER_TIMER1->TimerControl |= 0x80; //enable timer
+    uint32_t delta =  timestamp - us_ticker_read();
+    US_TICKER_TIMER2->TimerControl  &= ~CMSDK_DUALTIMER2_CTRL_EN_Msk;  // disable TIMER2
+    US_TICKER_TIMER2->TimerLoad = delta;                               // Set TIMER2 load value
+    US_TICKER_TIMER2->TimerControl |= CMSDK_DUALTIMER2_CTRL_INTEN_Msk; // enable TIMER2 interrupt
```

Comment:
I suggest to enable timer interrupts only if they are disabled.

Reply:
So you mean calling NVIC_EnableIRQ() before Enable TIMER2 counter?

Reply:
Something like:
```
if (US_TICKER_TIMER2->TimerControl & CMSDK_DUALTIMER2_CTRL_INTEN_Msk== 0) {
    US_TICKER_TIMER2->TimerControl |= CMSDK_DUALTIMER2_CTRL_INTEN_Msk;
    NVIC_EnableIRQ(US_TICKER_TIMER_IRQn);
}
```

Reply:
So is there any particular reasons for this?
what are the drawbacks for setting interrupt regardless of the current status? 

I was thinking about the case where an interrupt been set, but not timeout yet,  another interrupt got set again.  Or where NVIC got enabled for the TIMER_IRQ, but TimerControl interrupts BIT was not set.

Reply:
I was just thinking about optimisation (increase performance). `NVIC_EnableIRQ()` doesn't need to be called every time the interrupt is set.

Reply:
A write only operation would be probably faster than a "read/check" or "read/check/write".

Reply:
I think there is no guarantee that CMSDK_DUALTIMER2_CTRL_INTERUPT bit and NVIC_EnableIRQ() are in the same states (outside this driver code), so set NVIC_EnableIRQ() regardless CMSDK_DUALTIMER2_CTRL_INTERUPT would be "safer"?

Reply:
This is only suggestion and I didn't checked that. Maybe you are right, but please note there is an extra call to `NVIC_EnableIRQ() ` function. Additionally there is no point to do things which are already done. This function might be called very often when we have Ticker objects with small periods (`us`).